Index: insulate/insulate.py
===================================================================
--- insulate/insulate.py	(revision 38)
+++ insulate/insulate.py	(working copy)
@@ -39,6 +39,18 @@
     def getMethodName(test):
         return test._TestCase__testMethodName
 
+
+def getDropSlave( test ):
+    theTest = getattr( test , 'test' , None )  ## test is for example FunctionTestCase
+    if theTest:
+        drop = getattr( theTest , '__drop__' , False )
+    else:
+        drop = False
+    return drop
+
+
+
+
 class Insulate(Plugin):
     "Master insulation plugin class"
 
@@ -46,6 +58,9 @@
     name = 'insulate'
     restart_after_crash = True
     show_slave_output = False
+    discard_master_output = False
+    every_test = False
+    drop_slave = False
     enableOpt = 'insulate'
     testCount = 0
     testSlave = None
@@ -71,7 +86,22 @@
                           dest='show_slave_output',
                           action="store_true",
                           default=False)
-
+                          
+        parser.add_option('--insulate-discard-master-output',
+                          dest='discard_master_output',
+                          action="store_false",
+                          default=False)
+                          
+        parser.add_option('--insulate-every-test',
+                          dest='every_test',
+                          action="store_true",
+                          default=False)
+                          
+        parser.add_option('--insulate-drop-slave',
+                          dest='drop_slave',
+                          action="store_true",
+                          default=False)
+                           
     def configure(self, options, conf):
         super(Insulate, self).configure(options, conf)
         if getattr(options, 'insulateslave', None):
@@ -83,12 +113,32 @@
             self.in_slave = options.in_slave[:]
 
             self.show_slave_output = options.show_slave_output
+            self.discard_master_output = options.discard_master_output
+            self.every_test = options.every_test
+            self.drop_slave = options.drop_slave
 
             # Always enabled in slave
             self.in_slave.append('--nocapture')
 
         self.argv = [sys.executable] + sys.argv		# Save for slave creation in prepareTest
 
+
+    def dropSlave_(self):
+        if not(self.testSlave):return 
+        self.testSlave.dropSlave()
+        self.testSlave = None
+
+    def prepareSlave_(self):
+        args = [arg for arg in self.argv if arg not in self.not_in_slave]
+        args.extend(self.in_slave)
+        self.testSlave = TestSlave(self.restart_after_crash,
+                                   self.show_slave_output,
+                                   self.every_test,
+                                   self.drop_slave,
+                                   self.discard_master_output,
+                                   args)
+
+
     def prepareTest(self, test):
         """
         Prepare a test run; set counter to zero in preparation for run,
@@ -98,11 +148,7 @@
         Returns:   None
         """
         self.testCount = 0
-        args = [arg for arg in self.argv if arg not in self.not_in_slave]
-        args.extend(self.in_slave)
-        self.testSlave = TestSlave(self.restart_after_crash,
-                                   self.show_slave_output,
-                                   args)
+        self.prepareSlave_()
 
     def finalize(self, result):
         """
@@ -112,8 +158,7 @@
         Arguments: result
         Returns:   None
         """
-        self.testSlave.dropSlave()
-        self.testSlave = None
+        self.dropSlave_()
 
     def prepareTestCase(self, test):
         """
@@ -132,9 +177,14 @@
         Arguments: test - test to run
         Returns: result status
         """
+
+        if not(self.testSlave):
+            self.prepareSlave_()
+
         self.testCount += 1
 
         result = self.testSlave.runTest(self.testCount, test)
+
         return result
 
 class TestWrapper(object):
@@ -176,6 +226,8 @@
         method_name = getMethodName(self.orgtest)
         testMethod = getattr(self.orgtest, method_name)
 
+
+
         try:
             status, (ev, tb) = self.plugin.runTestInSlave(self.orgtest)
             if isinstance(ev, types.InstanceType):
@@ -201,7 +253,14 @@
 
         finally:
             result.stopTest(self.orgtest)
+            drop = getDropSlave(self.orgtest) 
+            if drop:
+                sys.stdout.write( ">>>>>>  dropping slave  after testMethod:[%s]  \n" % ( testMethod  ) )
+                self.plugin.dropSlave_()
+            else:
+                sys.stdout.write( ">>>>>>  NOT dropping slave  after testMethod:[%s] \n" % ( testMethod   ) )
 
+
 # Picklable objects mimicking a traceback structure enough
 class Code(object):
     def __init__(self, code):
@@ -230,13 +289,15 @@
 class TestSlave(object):
     "The test slave class for running tests in a separate process."
 
-    def __init__(self, restart_after_crash, show_slave_output, args):
+    def __init__(self, restart_after_crash, show_slave_output, every_test, drop_slave, discard_master_output, args):
         """
         Initialise the object, starting a 'slave' nosetest (or whatever
         the program was called) object in a separate process.
 
         Arguments: restart_after_crash - restart the slave after a crash
                    show_slave_output - show slave stdout, err
+                   every_test - retire the slave after every test 
+                   drop_slave - drop the slave after test when __drop__ is set to True on the test 
                    args - argument list
         Returns:   None
         """
@@ -244,6 +305,9 @@
 
         self.restart_after_crash = restart_after_crash
         self.show_slave_output = show_slave_output
+        self.discard_master_output = discard_master_output
+        self.every_test = every_test
+        self.drop_slave = drop_slave
         self.fromSlave = None
         self.toSlave = None
         self.noseSlave = None
@@ -291,6 +355,7 @@
         self.noseSlave = noseSlave
         self.toSlave = slavefile
         self.fromSlave = slavefile
+        sys.stdout.write(">>>>>>>>>>>> started slave on port %s  >>>>>>>>>>\n" % port )
 
     def dropSlave(self):
         """
@@ -309,6 +374,7 @@
             self.fromSlave.close()
             self.toSlave = self.fromSlave = None
         self.hasCrashed = True
+        
 
 
     def sendToSlave(self, count):
@@ -326,6 +392,7 @@
                           master and slave independently.
         Returns: (result status, result data)
         """
+
         self.startSlave()
 
         if self.noseSlave is None:
@@ -344,8 +411,15 @@
             if func_dict is not None:
                 orgtest.test.func_dict.update(func_dict)
             
-            sys.stdout.write(stdout)
-            sys.stderr.write(stderr)
+            if not self.discard_master_output:
+                sys.stdout.write(stdout)
+                sys.stderr.write(stderr)
+            else:
+                sys.stdout.write( ">>>>>>>>>>>>>>>>>>>>>>>>>  discarded master output" )
+            
+            if self.every_test:
+                sys.stdout.write( ">>>>>>>>>>>>>>>>>>>>>>>>>  dropping slave" )
+                self.dropSlave()
             return status, exc
 
         except (EOFError, IOError, socket.error), e:
