Index: trac/search/web_ui.py
===================================================================
--- trac/search/web_ui.py	(revision 7236)
+++ trac/search/web_ui.py	(working copy)
@@ -192,8 +192,8 @@
                 description = link.attrib.get('title', '')
         if quickjump_href:
             # Only automatically redirect to local quickjump links
-            if not quickjump_href.startswith(req.base_path or '/'):
-                noquickjump = True
+            #if not quickjump_href.startswith(req.base_path or '/'):
+            #    noquickjump = True
             if noquickjump:
                 return {'href': quickjump_href, 'name': tag.EM(name),
                         'description': description}
